ipn/store/kubestore,kube,envknob,cmd/tailscaled/depaware.txt: optionally run kubestore in cert share mode

irbekrm opened the pull request:
This PR adds some custom logic for reading and writing kube store values that are TLS certs and keys:

1) when store is initialized, look up additional TLS Secrets for this node and if found, load TLS certs from there

2) if the node runs in certs 'read only' mode and TLS cert and key are not found in the in-memory store, look those up in a Secret

3) if the node runs in certs 'read only' mode, run a daily TLS certs reload to memory to get any renewed certs

Updates tailscale/corp#24795

irbekrm commented on ipn/store/kubestore/store_kube.go:
```diff
@@ -40,13 +45,24 @@ type Store struct {
+	if err != nil {
+		return nil, err
+	}
+	return newWithClient(logf, c, secretName)
```
This split is for testing purposes (so I can pass in a mock client)

tomhjp commented on ipn/store/kubestore/store_kube.go:
```diff
@@ -202,6 +286,97 @@ func (s *Store) loadState() (err error) {
+// Currently (3/2025) this only applies to the Kubernetes Operator's ingress
+// ProxyGroup.
+func certSecretSelector() map[string]string {
+	podName := os.Getenv("POD_NAME")
```
We should move this into the 'new client' function and make its value a struct member.

tomhjp commented on ipn/store/kubestore/store_kube.go:
```diff
@@ -84,27 +116,79 @@ func (s *Store) ReadState(id ipn.StateKey) ([]byte, error) {
+	data := map[string][]byte{domain + ".crt": cert, domain + ".key": key}
+	// If we run in cert share mode, cert and key for a DNS name are written
+	// to a separate Secret.
+	if envknob.IsCertShareReadWriteMode() {
```
Should we return an error if we're in readonly mode?

irbekrm replied:
Theoretically, that would be the correct thing to do, but in practice if further down the line someone uses it in some sort of custom setup and misconfigures it so that a cert and key actually got issued in readonly mode, I'd rather store it and lessen the chance of them hitting the limits than error out because it was incorrect to issue in this mode and throw the cert away. I can add a log line

tomhjp commented on ipn/store/kubestore/store_kube.go:
```diff
@@ -84,27 +116,79 @@ func (s *Store) ReadState(id ipn.StateKey) ([]byte, error) {
+func (s *Store) WriteTLSCertAndKey(domain string, cert, key []byte) (err error) {
+	defer func() {
+		if err == nil {
+			s.memory.WriteState(ipn.StateKey(sanitizeKey(domain+".crt")), cert)
+			s.memory.WriteState(ipn.StateKey(sanitizeKey(domain+".key")), key)
```
If we get very unlucky, we could race with the periodic refresh goroutine. I think it's worth having a mutex around the whole of the `WriteTLSCertAndKey` and `ReadTLSCertAndKey` functions, and the relevant part of `loadCerts` just to make sure the Secret is always consistent with the in-memory store, and the cert is always consistent with the key.

irbekrm replied:
I think what I want to do is to have

```
func (s *Store) ReadStates(ids []ipn.StateKey) (map[ipn.StateKey][]byte, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
...
```
and similar for Write and call those from `WriteTLSCertAndKey`, `ReadTLSCertAndKey`.
I'm uneasy about doing a thing where we add a lock to the kube store given that we also lock the inner memory store- I think that could end up with deadlocks especially when the other `ReadState`, `WriteState` method will only lock the inner struct.

But then I'd rather put this Store interface change into a small separate PR

irbekrm replied:
I just added a TODO for now

tomhjp commented on ipn/store/kubestore/store_kube.go:
```diff
@@ -84,27 +116,79 @@ func (s *Store) ReadState(id ipn.StateKey) ([]byte, error) {
+			return cert, key, nil
+		}
+	}
+	if !envknob.IsCertShareReadOnlyMode() {
```
It's not obvious to me on first read why we don't return early if we're in read/write mode. Maybe worth a code comment?

tomhjp commented on ipn/store/kubestore/store_kube.go:
```diff
@@ -84,27 +116,79 @@ func (s *Store) ReadState(id ipn.StateKey) ([]byte, error) {
+// memory and running cert store in read-only mode, looks up a Secret.
+func (s *Store) ReadTLSCertAndKey(domain string) (cert, key []byte, err error) {
+	// Try memory first - use sanitized keys
+	certKey := sanitizeKey(domain + ".crt")
```
Given that a Kubernetes Secret name must be a valid DNS name, can we instead error if the domain needs sanitising? I think it would make the cert state keys a bit easier to reason about knowing for sure that we don't need to worry about overlaps due to mutations.

tomhjp commented on ipn/store/kubestore/store_kube.go:
```diff
@@ -84,27 +116,79 @@ func (s *Store) ReadState(id ipn.StateKey) ([]byte, error) {
+func (s *Store) WriteTLSCertAndKey(domain string, cert, key []byte) (err error) {
+	defer func() {
+		if err == nil {
+			s.memory.WriteState(ipn.StateKey(sanitizeKey(domain+".crt")), cert)
```
WDYT about using a separate in-memory datastore for certs? We are overlapping the keyspace of multiple Secrets here, which feels like it would make it easy to introduce subtle bugs in future.

irbekrm replied:
I was considering that for the previous re-incarnation of the cert sharing where all the certs for ProxyGroup would have come from one Secret and it would been a clear mapping between two stores and two Secrets.
I think at this point, I'd rather avoid changing even more internals unless there is a clear benefit- we should see better how to structure all this once we have BYO certs and perhaps once users start needing to use the read only mode for their own VIPService certs, if they do

tomhjp reviewed (commented):
Looks good 👍 I found it a bit hard to reason about some of the edge cases, so some of my comments are just looking for opportunities to simplify

irbekrm commented:
Thanks @tomhjp I believe I have addressed all the comments PTAL (I had to merge them into the same single commit as a result of some unfortunate git actions)

tomhjp commented on ipn/store/kubestore/store_kube.go:
```diff
@@ -84,27 +126,99 @@ func (s *Store) ReadState(id ipn.StateKey) ([]byte, error) {
+		}
+	}
+	if s.certShareMode != "ro" {
+		return nil, nil, ipn.ErrStateNotExist
```
Looking at the call sites, this is a bigger change for another PR, but it would be good to use `errors.Is` to look for this sentinel error so that we can include more information about _why_ the cert doesn't exist. I could see it being tricky to debug without getting an interactive session in the cluster.

irbekrm replied:
Agree, added a TODO

tomhjp reviewed (approved):
Thanks! LGTM 👍
